\documentclass{article}
\usepackage{fullpage}
\usepackage{hyperref}

%\VignetteIndexEntry{Using RQT, an R/Bioconductor package for gene-level meta-analysis}

\title{Using RQT, an R/Bioconductor package for gene-level meta-analysis}
\author{Ilya Y. Zhbannikov}
\date{\today}


\begin{document}
\maketitle

\section{Overview}

Despite the recent advances of modern GWAS methods, 
it is still remains an important problem of addressing 
calculation an effect size and corresponding p-value 
for the whole gene rather than for single variant. 
We developed an R-package rqt, which offers gene-level GWAS meta-analysis. 
The package can be easily included into bioinformatics pipeline 
or used as stand-alone. 
Contact: ilya.zhbannikov@duke.edu for questions of 
usage the \texttt{rqt} or any other issues.

Below we provide several examples that show GWAS 
meta-analysis on gene-level layer.

\subsection{Methods in brief}

The workflow of gene-level meta analysis consists of the following steps: 
(i) reducing the number of predictors, thereby alleviating 
correlation problem in variants (accounting for LD); 
(ii) then the regression mod-el is fitted on the reduced dataset 
to obtain corresponding regression coefficient ("effect sizes"); 
(iii) these coefficients are then to be pooled into a total index 
representing a total gene-level effect size and corresponding 
statistics is calculated. P- and q- values are then calculated 
using this statistics from asymptotic approximation or permutation 
procedure; (iv) the final step is combining gene-level p-values 
calculated from each study with Fisher's combined probability method.

\section{Installation}

\subsection{Most-recent version from GitHub}

<<eval=FALSE>>=
require(devtools)
devtools::install_github("izhbannikov/rqt")
@

\section{Data description}
\subsection{Single dataset}
In \texttt{rqt} requires the following datasets: 
(i) \texttt{phenotype} (a \texttt{N} by 1) matrix (i.e. a vector); 
and (ii) \texttt{genotype} - an object of class 
\texttt{SummarizedExperiment} containing one assay:
(a \texttt{N} by \texttt{M}) matrix, where 
\texttt{N} - is the total 
number of individuals in the study and \texttt{M} is the total number 
of genetic variants. Optionally, \texttt{rqt} can accept covariates, 
in form of \texttt{N} by \texttt{K} matrix, where \texttt{K} 
is the total number of 
covariates used in the study. Phenotype can be dichotomous 
(0/1, where 1 indicates control and 0 case).

\subsection{Meta-analysis}
In meta-analysis, \texttt{rqt} requires a list of \texttt{M} 
(\texttt{M} - number 
of datasets used in meta-analysis) and optionally it accepts
covariates in form described above.

\section{Examples}

\subsection{Gene-level analysis on a single dataset}
\subsection{Dichotomous phenotype}

<<>>=
library(rqt)
data <- data.matrix(read.table(system.file("extdata/test.bin1.dat",
                                           package="rqt"), header=TRUE))
pheno <- data[,1]
geno <- data[, 2:dim(data)[2]]
colnames(geno) <- paste(seq(1, dim(geno)[2]))
geno.obj <- SummarizedExperiment(geno)
obj <- rqtClass(phenotype=pheno, genotype=geno.obj)
res <- rQTest(obj, method="pca", out.type = "D")
print(res)
@

\subsection{Continuous phenotype}

<<>>=
library(rqt)
data <- data.matrix(read.table(system.file("extdata/test.cont1.dat",
                                           package="rqt"), header = TRUE))
pheno <- data[,1]
geno <- data[, 2:dim(data)[2]]
colnames(geno) <- paste(seq(1, dim(geno)[2]))
geno.obj <- SummarizedExperiment(geno)
obj <- rqtClass(phenotype=pheno, genotype=geno.obj)
res <- rQTest(obj, method="pca", out.type = "C")
print(res)
@

\subsection{Preprocessing with Partial Least Square regression (PLS)}

This method is used for continous outcome, i.e. out.type = "C".
<<>>=
library(rqt)
data <- data.matrix(read.table(system.file("extdata/test.cont1.dat",
                                           package="rqt"), header = TRUE))
pheno <- data[,1]
geno <- data[, 2:dim(data)[2]]
colnames(geno) <- paste(seq(1, dim(geno)[2]))
geno.obj <- SummarizedExperiment(geno)
obj <- rqtClass(phenotype=pheno, genotype=geno.obj)
res <- rQTest(obj, method="pls", out.type = "C")
print(res)
@

\subsection{Preprocessing with Partial Least Square Discriminant Analysis (PLS-DA)}

This method of data preprocessing is used for dichotomous outcome.

<<>>=
library(rqt)
data <- data.matrix(read.table(system.file("extdata/test.bin1.dat",
                                           package="rqt"), header=TRUE))
pheno <- data[,1]
geno <- data[, 2:dim(data)[2]]
colnames(geno) <- paste(seq(1, dim(geno)[2]))
geno.obj <- SummarizedExperiment(geno)
obj <- rqtClass(phenotype=pheno, genotype=geno.obj)
res <- rQTest(obj, method="pls", out.type = "D", scale = TRUE)
print(res)
@

\subsection{Using additional covariates}

Quite often, researchers want to supply not only genetic 
data but also specific covariates, 
representic some physiological parameters or environment 
(for example, to evaluate 
hyphoteses of gene-environment interactions). 
In such cases, the package \texttt{rqt} 
can accept additional covariates, in form of 
\texttt{N} by \texttt{K} matrix, as provided below:

<<>>=
library(rqt)
data <- data.matrix(read.table(system.file("extdata/test.bin1.dat",
                                           package="rqt"), header = TRUE))
pheno <- data[,1]
geno <- data[, 2:dim(data)[2]]
colnames(geno) <- paste(seq(1, dim(geno)[2]))
geno.obj <- SummarizedExperiment(geno)
covars <- read.table(system.file("extdata/test.cova1.dat",package="rqt"), 
    header=TRUE)
obj <- rqtClass(phenotype=pheno, genotype=geno.obj, covariates = covars)
res <- rQTest(obj, method="pca", out.type = "D")
print(res)
@


\section{Meta-analysis}

<<>>=
library(rqt)
data1 <- data.matrix(read.table(system.file("extdata/phengen2.dat",
                                            package="rqt"), skip=1))
pheno <- data1[,1]
geno <- data1[, 2:dim(data1)[2]]
colnames(geno) <- paste(seq(1, dim(geno)[2]))
geno.obj <- SummarizedExperiment(geno)
obj1 <- rqtClass(phenotype=pheno, genotype=geno.obj)

data2 <- data.matrix(read.table(system.file("extdata/phengen3.dat",
                                            package="rqt"), skip=1))
pheno <- data2[,1]
geno <- data2[, 2:dim(data2)[2]]
colnames(geno) <- paste(seq(1, dim(geno)[2]))
geno.obj <- SummarizedExperiment(geno)
obj2 <- rqtClass(phenotype=pheno, genotype=geno.obj)

data3 <- data.matrix(read.table(system.file("extdata/phengen.dat",
                                            package="rqt"), skip=1))
pheno <- data3[,1]
geno <- data3[, 2:dim(data3)[2]]
colnames(geno) <- paste(seq(1, dim(geno)[2]))
geno.obj <- SummarizedExperiment(geno)
obj3 <- rqtClass(phenotype=pheno, genotype=geno.obj)

res.meta <- rQTestMeta(list(obj1, obj2, obj3))
print(res.meta)
@


\section{Session information}
<<echo=TRUE, results=verbatim>>=
sessionInfo()
@
%--------
\clearpage
\section{References}
\begin{thebibliography}{00}

\bibitem{Wu2011} Wu, M. C., Lee, S., Cai, T., Li, Y., Boehnke, M., 
and Lin, X. (2011) 
Rare Variant Association Testing for Sequencing Data Using the Sequence 
Kernel Associa-tion Test (SKAT). The American Journal of 
Human Genetics, 89, 82-93.
\bibitem{Li2008} Li, B. and Leal, S.M. (2008) Methods for 
Detecting Associations 
with Rare Vari-ants for Common Diseases: Application to Analysis 
of Sequence Data. The American Journal of Human Genetics, 83, 311-321.
\bibitem{Liu2010} Liu D., Leal S. (2010) A Novel Adaptive Method for 
the Analysis 
of Next-Generation Sequencing Data to Detect Complex Trait 
Associations with Rare Variants Due to Gene Main Effects and 
Interactions, PLoS Genet., 6(10).
\bibitem{Madsen2009} Madsen, B.E, Browning, S.R. (2009) 
A Groupwise Association 
Test for Rare Mutations Using a Weighted Sum Statistic, PLoS Genet., 5(2).

\bibitem{Lee2016} Lee, J., Kim, Y.J., Lee, J., T2D-Genes Consortium, 
Kim, B-J., 
Lee, S., Park T. (2016) Gene-set association tests for 
next-generation sequencing data, Bioinformatics, 32(17).

\bibitem{Tibshirani1996} Tibshirani, R. (1996) Regression shrinkage 
and selection 
via the lasso. J. Royal. Statist. Soc B., Vol. 58, No. 1, pages 267-288.
\end{thebibliography}
%--------

\end{document}