---
title: "rqt: an R/Bioconductor package for gene-level meta-analysis"
author: "Ilya Zhbannikov"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---

## Overview

Despite the recent advances of modern GWAS methods, it is still remains an important problem of addressing calculation an effect size and corresponding p-value for the whole gene rather than for single variant. We developed an R-package rqt, which offers gene-level GWAS meta-analysis. The package can be easily included into bioinformatics pipeline or used as stand-alone. 
Contact: ilya.zhbannikov@duke.edu for questions of usage the \code{rqt} or any other issues.

Below we provide several examples that show GWAS meta-analysis on gene-level layer.

### Methods in brief

The workflow of gene-level meta analysis consists of the following steps: (i) reducing the number of predictors, thereby alleviating correlations between variants (accounting for LD); (ii) then the regression mod-el is fitted on the reduced dataset to obtain corresponding regression coefficient ("effect sizes"); (iii) these coefficients are then to be pooled into a total index representing a total gene-level effect size and corresponding statistics is calculated. P- and q- values are then calculated using this statistics from asymptotic approximation or permutation procedure; (iv) the final step is combining gene-level p-values calculated from each study with Fisher's combined probability method.

## Installation
### Stable version from Bioconductor

TODO

### Most-recent version from GitHub
```{r,eval=FALSE}
require(devtools)
devtools::install_github("izhbannikov/rqt")
```


## Data description
### Single dataset ###
In \code{rqt} requires two datasets: phenotype (a n by 1) matrix and genotype (a n by m) matrix, where n - is the total 
number of individuals in the study and m is the total number of genetic variants. Optionally, rqt can accept covariates, 
in form of n by k matrix, where k is the total number of covariates used in the study. Phenotype can be dichotomous 
(0/1, where 1 indicates control and 0 case).

### Meta-analysis ###
In meta-analysis, \code{rqt} requires a list of M (M - number of datasets used in meta-analysis) and optionally it accepts
covariates in form described above.

## Examples

### Gene-level analysis on a single dataset
### Dichotomous phenotype
```{r}
library(rqt)
data <- read.table(system.file("data/test.bin1.dat",package="rqt"), header = TRUE)
pheno <- data$pheno
geno <- data[, 2:dim(data)[2]]
obj <- rqtClass(phenotype=pheno, genotype=geno)
res <- rQTest(obj, method="pca", out.type = "D")
print(res)
```

### Continuous phenotype
```{r}
library(rqt)
data <- read.table(system.file("data/test.cont1.dat",package="rqt"), header = TRUE)
pheno <- data$pheno
geno <- data[, 2:dim(data)[2]]
obj <- rqtClass(phenotype=pheno, genotype=geno)
res <- rQTest(obj, method="pca", out.type = "C")
print(res)
```

### Preprocessing with Partial Least Square regression (PLS)

This method is used for continous outcome, i.e. out.type = "C".
```{r}
library(rqt)
data <- read.table(system.file("data/test.cont1.dat",package="rqt"), header = TRUE)
pheno <- data$pheno
geno <- data[, 2:dim(data)[2]]
obj <- rqtClass(phenotype=pheno, genotype=geno)
res <- rQTest(obj, method="pls", out.type = "C")
print(res)
```

### Preprocessing with Partial Least Square Discriminant Analysis (PLS-DA)

This method of data preprocessing is used for dichotomous outcome.
```{r}
library(rqt)
data <- read.table(system.file("data/test.bin1.dat",package="rqt"), header=TRUE)
pheno <- data$pheno
geno <- data[, 2:dim(data)[2]]
obj <- rqtClass(phenotype=pheno, genotype=geno)
res <- rQTest(obj, method="pls", out.type = "D", scale = TRUE)
print(res)
```

### Using additional covariates

Quite often, researchers want to supply not only genetic data but also specific covariates, 
representic some physiological parameters or environment (for example, to evaluate 
hyphoteses of gene-environment interactions). In such cases, the package \code{rqt} 
can accept additional covariates, in form of n by k matrix, as provided below:

```{r}
library(rqt)
data <- read.table(system.file("data/test.bin1.dat",package="rqt"), header = TRUE)
pheno <- data$pheno
geno <- data[, 2:dim(data)[2]]
covars <- read.table(system.file("data/test.cova1.dat",package="rqt"), header=TRUE)
obj <- rqtClass(phenotype=pheno, genotype=geno, covariates = covars)
res <- rQTest(obj, method="pca", out.type = "D")
print(res)
```

### Survival analysis

TODO


## Meta-analysis

```{r}
library(rqt)
data1 <- read.table(system.file("data/phengen2.dat",package="rqt"), skip=1)
obj1 <- rqtClass(phenotype=data1[,1], genotype=data1[, 2:dim(data1)[2]])
data2 <- read.table(system.file("data/phengen3.dat",package="rqt"), skip=1)
obj2 <- rqtClass(phenotype=data2[,1], genotype=data2[, 2:dim(data2)[2]])
data3 <- read.table(system.file("data/phengen.dat",package="rqt"), skip=1)
obj3 <- rqtClass(phenotype=data3[,1], genotype=data3[, 2:dim(data3)[2]])
res.meta <- rQTestMeta(list(obj1, obj2, obj3))
print(res.meta)
```

## References

* Wu, M. C., Lee, S., Cai, T., Li, Y., Boehnke, M., and Lin, X. (2011) Rare Variant Association Testing for Sequencing Data Using the Sequence Kernel Associa-tion Test (SKAT). The American Journal of Human Genetics, 89, 82-93.

* Li, B. and Leal, S.M. (2008) Methods for Detecting Associations with Rare Vari-ants for Common Diseases: Application to Analysis of Sequence Data. The American Journal of Human Genetics, 83, 311-321.

* Liu D., Leal S. (2010) A Novel Adaptive Method for the Analysis of Next-Generation Sequencing Data to Detect Complex Trait Associations with Rare Variants Due to Gene Main Effects and Interactions, PLoS Genet., 6(10).

* Madsen, B.E, Browning, S.R. (2009) A Groupwise Association Test for Rare Mutations Using a Weighted Sum Statistic, PLoS Genet., 5(2).

* Lee, J., Kim, Y.J., Lee, J., T2D-Genes Consortium, Kim, B-J., Lee, S., Park T. (2016) Gene-set association tests for next-generation sequencing data, Bioinformatics, 32(17).

* Tibshirani, R. (1996) Regression shrinkage and selection via the lasso. J. Royal. Statist. Soc B., Vol. 58, No. 1, pages 267-288.
