\documentclass{article}
\usepackage{fullpage}
\usepackage{hyperref}

%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{Using RQT, an R/Bioconductor package for gene-level meta-analysis}

\title{Using RQT, an R/Bioconductor package for gene-level meta-analysis}
\author{Ilya Y. Zhbannikov}
\date{\today}


\usepackage{Sweave}
\begin{document}
\input{rqt-vignette-concordance}
\maketitle

\section{Overview}

Despite the recent advances of modern GWAS methods, 
it is still remains an important problem of addressing 
calculation an effect size and corresponding p-value 
for the whole gene rather than for single variant. 
We developed an R-package rqt, which offers gene-level GWAS meta-analysis. 
The package can be easily included into bioinformatics pipeline 
or used as stand-alone. 
Contact: ilya.zhbannikov@duke.edu for questions of 
usage the \texttt{rqt} or any other issues.

Below we provide several examples that show GWAS 
meta-analysis on gene-level layer.

\subsection{Methods in brief}

The workflow of gene-level meta analysis consists of the following steps: 
(i) reducing the number of predictors, thereby alleviating 
correlation problem in variants (accounting for LD); 
(ii) then the regression mod-el is fitted on the reduced dataset 
to obtain corresponding regression coefficient ("effect sizes"); 
(iii) these coefficients are then to be pooled into a total index 
representing a total gene-level effect size and corresponding 
statistics is calculated. P- and q- values are then calculated 
using this statistics from asymptotic approximation or permutation 
procedure; (iv) the final step is combining gene-level p-values 
calculated from each study with Fisher's combined probability method.

\section{Installation}

\subsection{Most-recent version from GitHub}

\begin{Schunk}
\begin{Sinput}
> require(devtools)
> devtools::install_github("izhbannikov/rqt")
\end{Sinput}
\end{Schunk}

\section{Data description}
\subsection{Single dataset}
In \texttt{rqt} requires two datasets: phenotype (a n by 1) matrix 
and genotype (a n by m) matrix, where n - is the total 
number of individuals in the study and m is the total number 
of genetic variants. Optionally, rqt can accept covariates, 
in form of n by k matrix, where k is the total number of 
covariates used in the study. Phenotype can be dichotomous 
(0/1, where 1 indicates control and 0 case).

\subsection{Meta-analysis}
In meta-analysis, \texttt{rqt} requires a list of M (M - number 
of datasets used in meta-analysis) and optionally it accepts
covariates in form described above.

\section{Examples}

\subsection{Gene-level analysis on a single dataset}
\subsection{Dichotomous phenotype}

\begin{Schunk}
\begin{Sinput}
> library(rqt)
> data <- read.table(system.file("extdata/test.bin1.dat",package="rqt"), 
+     header = TRUE)
> pheno <- data$pheno
> geno <- data[, 2:dim(data)[2]]
> obj <- rqtClass(phenotype=pheno, genotype=geno)
> res <- rQTest(obj, method="pca", out.type = "D")
> print(res)
\end{Sinput}
\begin{Soutput}
Phenotype:
[1] 1 1 1 1 1 1
...

Genotype:
  snp1 snp2 snp3 snp4 snp5 snp6 snp7 snp8 snp9 snp10 snp11 snp12 snp13 snp14
1    0    0    0    1    0    2    1    0    0     0     2     1     2     0
2    1    0    1    0    0    1    0    0    0     0     0     0     1     0
3    0    0    0    0    1    0    0    1    0     1     1     1     0     0
4    0    0    1    0    1    0    0    1    1     0     0     0     1     0
5    0    0    1    1    1    1    1    1    0     1     0     1     0     0
6    0    0    1    1    1    0    0    1    0     1     1     0     1     0
  snp15 snp16 snp17 snp18 snp19 snp20 snp21 snp22 snp23 snp24 snp25 snp26 snp27
1     0     2     1     1     0     0     1     0     0     0     0     2     1
2     0     0     2     1     2     1     0     1     0     0     0     2     1
3     0     1     1     0     0     1     1     1     0     0     0     1     0
4     0     0     1     0     0     2     0     1     0     0     0     1     0
5     0     0     0     0     0     0     1     0     0     1     0     1     2
6     0     2     0     0     1     0     1     1     0     0     0     2     0
  snp28 snp29 snp30 snp31 snp32 snp33 snp34 snp35 snp36 snp37 snp38 snp39 snp40
1     0     1     0     0     0     0     0     0     1     1     0     0     1
2     1     2     0     0     0     1     2     1     2     0     1     1     0
3     0     0     0     0     0     0     1     1     1     2     2     0     0
4     2     0     0     0     1     2     2     0     1     1     1     1     0
5     0     0     0     0     0     1     1     0     0     1     1     0     0
6     0     0     0     0     0     1     0     1     1     0     0     2     0
  snp41 snp42 snp43 snp44 snp45 snp46 snp47 snp48 snp49 snp50 snp51 snp52 snp53
1     0     0     1     0     0     0     0     1     1     0     0     0     1
2     1     0     0     1     0     0     0     2     0     1     0     1     0
3     0     1     0     1     0     2     1     1     1     0     0     0     1
4     0     0     1     1     0     0     1     1     2     0     1     0     2
5     2     2     0     0     0     1     2     1     0     0     0     0     1
6     1     1     2     0     0     0     1     1     1     0     0     1     1
  snp54 snp55 snp56 snp57 snp58 snp59 snp60 snp61 snp62 snp63 snp64 snp65 snp66
1     1     1     1     1     0     0     0     0     1     1     0     1     1
2     2     1     0     1     0     0     0     1     0     1     2     2     1
3     2     1     0     0     0     0     0     0     0     1     0     0     1
4     1     1     0     0     0     0     2     1     0     1     0     0     2
5     1     1     1     0     0     0     2     1     1     0     0     1     1
6     0     1     1     0     0     0     0     0     0     0     1     2     1
  snp67 snp68 snp69 snp70 snp71 snp72 snp73 snp74 snp75 snp76 snp77 snp78 snp79
1     1     0     1     1     2     1     1     1     0     0     2     1     1
2     1     0     1     0     1     0     0     0     0     1     1     2     1
3     1     0     1     2     1     0     1     1     0     0     0     0     1
4     1     0     0     1     1     1     0     0     0     0     0     1     0
5     0     0     1     1     0     0     1     0     1     0     0     1     0
6     1     0     0     0     0     0     1     1     1     1     1     1     2
  snp80 snp81 snp82 snp83 snp84 snp85 snp86 snp87 snp88 snp89 snp90 snp91 snp92
1     1     0     1     1     0     1     1     0     0     0     1     0     1
2     1     2     0     0     0     1     2     0     2     1     0     0     2
3     1     0     0     0     0     2     0     0     2     0     0     0     0
4     1     2     0     1     0     1     2     0     0     0     0     0     0
5     2     0     1     0     0     2     0     0     0     0     1     0     0
6     1     0     0     1     0     0     0     0     2     0     1     0     0
  snp93 snp94 snp95 snp96 snp97 snp98 snp99 snp100
1     1     0     0     0     1     0     1      1
2     1     0     0     0     0     1     1      2
3     1     1     0     0     1     2     0      0
4     1     0     0     0     1     0     0      0
5     1     0     0     1     1     2     1      1
6     0     0     0     1     1     0     1      1
...

Covariates:
data frame with 0 columns and 0 rows


Results:

$Qstatistic
         Q1        Q2        Q3
1 0.9086994 0.9938212 0.9086994

$p.value
       p.Q1      p.Q2      p.Q3
1 0.3404597 0.9285409 0.5654533
\end{Soutput}
\end{Schunk}

\subsection{Continuous phenotype}

\begin{Schunk}
\begin{Sinput}
> library(rqt)
> data <- read.table(system.file("extdata/test.cont1.dat",package="rqt"), 
+     header = TRUE)
> pheno <- data$pheno
> geno <- data[, 2:dim(data)[2]]
> obj <- rqtClass(phenotype=pheno, genotype=geno)
> res <- rQTest(obj, method="pca", out.type = "C")
> print(res)
\end{Sinput}
\begin{Soutput}
Phenotype:
[1] 3.422452 2.457394 2.708564 4.589394 5.461723 4.438707
...

Genotype:
  snp1 snp2 snp3 snp4 snp5 snp6 snp7 snp8 snp9 snp10 snp11 snp12 snp13 snp14
1    0    0    0    1    0    2    1    0    0     0     2     1     2     0
2    1    0    1    0    0    1    0    0    0     0     0     0     1     0
3    0    0    0    0    1    0    0    1    0     1     1     1     0     0
4    0    0    1    0    1    0    0    1    1     0     0     0     1     0
5    0    0    1    1    1    1    1    1    0     1     0     1     0     0
6    0    0    1    1    1    0    0    1    0     1     1     0     1     0
  snp15 snp16 snp17 snp18 snp19 snp20 snp21 snp22 snp23 snp24 snp25 snp26 snp27
1     0     2     1     1     0     0     1     0     0     0     0     2     1
2     0     0     2     1     2     1     0     1     0     0     0     2     1
3     0     1     1     0     0     1     1     1     0     0     0     1     0
4     0     0     1     0     0     2     0     1     0     0     0     1     0
5     0     0     0     0     0     0     1     0     0     1     0     1     2
6     0     2     0     0     1     0     1     1     0     0     0     2     0
  snp28 snp29 snp30 snp31 snp32 snp33 snp34 snp35 snp36 snp37 snp38 snp39 snp40
1     0     1     0     0     0     0     0     0     1     1     0     0     1
2     1     2     0     0     0     1     2     1     2     0     1     1     0
3     0     0     0     0     0     0     1     1     1     2     2     0     0
4     2     0     0     0     1     2     2     0     1     1     1     1     0
5     0     0     0     0     0     1     1     0     0     1     1     0     0
6     0     0     0     0     0     1     0     1     1     0     0     2     0
  snp41 snp42 snp43 snp44 snp45 snp46 snp47 snp48 snp49 snp50 snp51 snp52 snp53
1     0     0     1     0     0     0     0     1     1     0     0     0     1
2     1     0     0     1     0     0     0     2     0     1     0     1     0
3     0     1     0     1     0     2     1     1     1     0     0     0     1
4     0     0     1     1     0     0     1     1     2     0     1     0     2
5     2     2     0     0     0     1     2     1     0     0     0     0     1
6     1     1     2     0     0     0     1     1     1     0     0     1     1
  snp54 snp55 snp56 snp57 snp58 snp59 snp60 snp61 snp62 snp63 snp64 snp65 snp66
1     1     1     1     1     0     0     0     0     1     1     0     1     1
2     2     1     0     1     0     0     0     1     0     1     2     2     1
3     2     1     0     0     0     0     0     0     0     1     0     0     1
4     1     1     0     0     0     0     2     1     0     1     0     0     2
5     1     1     1     0     0     0     2     1     1     0     0     1     1
6     0     1     1     0     0     0     0     0     0     0     1     2     1
  snp67 snp68 snp69 snp70 snp71 snp72 snp73 snp74 snp75 snp76 snp77 snp78 snp79
1     1     0     1     1     2     1     1     1     0     0     2     1     1
2     1     0     1     0     1     0     0     0     0     1     1     2     1
3     1     0     1     2     1     0     1     1     0     0     0     0     1
4     1     0     0     1     1     1     0     0     0     0     0     1     0
5     0     0     1     1     0     0     1     0     1     0     0     1     0
6     1     0     0     0     0     0     1     1     1     1     1     1     2
  snp80 snp81 snp82 snp83 snp84 snp85 snp86 snp87 snp88 snp89 snp90 snp91 snp92
1     1     0     1     1     0     1     1     0     0     0     1     0     1
2     1     2     0     0     0     1     2     0     2     1     0     0     2
3     1     0     0     0     0     2     0     0     2     0     0     0     0
4     1     2     0     1     0     1     2     0     0     0     0     0     0
5     2     0     1     0     0     2     0     0     0     0     1     0     0
6     1     0     0     1     0     0     0     0     2     0     1     0     0
  snp93 snp94 snp95 snp96 snp97 snp98 snp99 snp100
1     1     0     0     0     1     0     1      1
2     1     0     0     0     0     1     1      2
3     1     1     0     0     1     2     0      0
4     1     0     0     0     1     0     0      0
5     1     0     0     1     1     2     1      1
6     0     0     0     1     1     0     1      1
...

Covariates:
data frame with 0 columns and 0 rows


Results:

$Qstatistic
         Q1       Q2       Q3
1 0.2846585 2.389537 2.219594

$p.value
      p.Q1      p.Q2      p.Q3
1 0.593664 0.7022459 0.2561996
\end{Soutput}
\end{Schunk}

\subsection{Preprocessing with Partial Least Square regression (PLS)}

This method is used for continous outcome, i.e. out.type = "C".
\begin{Schunk}
\begin{Sinput}
> library(rqt)
> data <- read.table(system.file("extdata/test.cont1.dat",package="rqt"), 
+     header = TRUE)
> pheno <- data$pheno
> geno <- data[, 2:dim(data)[2]]
> obj <- rqtClass(phenotype=pheno, genotype=geno)
> res <- rQTest(obj, method="pls", out.type = "C")
> print(res)
\end{Sinput}
\begin{Soutput}
Phenotype:
[1] 3.422452 2.457394 2.708564 4.589394 5.461723 4.438707
...

Genotype:
  snp1 snp2 snp3 snp4 snp5 snp6 snp7 snp8 snp9 snp10 snp11 snp12 snp13 snp14
1    0    0    0    1    0    2    1    0    0     0     2     1     2     0
2    1    0    1    0    0    1    0    0    0     0     0     0     1     0
3    0    0    0    0    1    0    0    1    0     1     1     1     0     0
4    0    0    1    0    1    0    0    1    1     0     0     0     1     0
5    0    0    1    1    1    1    1    1    0     1     0     1     0     0
6    0    0    1    1    1    0    0    1    0     1     1     0     1     0
  snp15 snp16 snp17 snp18 snp19 snp20 snp21 snp22 snp23 snp24 snp25 snp26 snp27
1     0     2     1     1     0     0     1     0     0     0     0     2     1
2     0     0     2     1     2     1     0     1     0     0     0     2     1
3     0     1     1     0     0     1     1     1     0     0     0     1     0
4     0     0     1     0     0     2     0     1     0     0     0     1     0
5     0     0     0     0     0     0     1     0     0     1     0     1     2
6     0     2     0     0     1     0     1     1     0     0     0     2     0
  snp28 snp29 snp30 snp31 snp32 snp33 snp34 snp35 snp36 snp37 snp38 snp39 snp40
1     0     1     0     0     0     0     0     0     1     1     0     0     1
2     1     2     0     0     0     1     2     1     2     0     1     1     0
3     0     0     0     0     0     0     1     1     1     2     2     0     0
4     2     0     0     0     1     2     2     0     1     1     1     1     0
5     0     0     0     0     0     1     1     0     0     1     1     0     0
6     0     0     0     0     0     1     0     1     1     0     0     2     0
  snp41 snp42 snp43 snp44 snp45 snp46 snp47 snp48 snp49 snp50 snp51 snp52 snp53
1     0     0     1     0     0     0     0     1     1     0     0     0     1
2     1     0     0     1     0     0     0     2     0     1     0     1     0
3     0     1     0     1     0     2     1     1     1     0     0     0     1
4     0     0     1     1     0     0     1     1     2     0     1     0     2
5     2     2     0     0     0     1     2     1     0     0     0     0     1
6     1     1     2     0     0     0     1     1     1     0     0     1     1
  snp54 snp55 snp56 snp57 snp58 snp59 snp60 snp61 snp62 snp63 snp64 snp65 snp66
1     1     1     1     1     0     0     0     0     1     1     0     1     1
2     2     1     0     1     0     0     0     1     0     1     2     2     1
3     2     1     0     0     0     0     0     0     0     1     0     0     1
4     1     1     0     0     0     0     2     1     0     1     0     0     2
5     1     1     1     0     0     0     2     1     1     0     0     1     1
6     0     1     1     0     0     0     0     0     0     0     1     2     1
  snp67 snp68 snp69 snp70 snp71 snp72 snp73 snp74 snp75 snp76 snp77 snp78 snp79
1     1     0     1     1     2     1     1     1     0     0     2     1     1
2     1     0     1     0     1     0     0     0     0     1     1     2     1
3     1     0     1     2     1     0     1     1     0     0     0     0     1
4     1     0     0     1     1     1     0     0     0     0     0     1     0
5     0     0     1     1     0     0     1     0     1     0     0     1     0
6     1     0     0     0     0     0     1     1     1     1     1     1     2
  snp80 snp81 snp82 snp83 snp84 snp85 snp86 snp87 snp88 snp89 snp90 snp91 snp92
1     1     0     1     1     0     1     1     0     0     0     1     0     1
2     1     2     0     0     0     1     2     0     2     1     0     0     2
3     1     0     0     0     0     2     0     0     2     0     0     0     0
4     1     2     0     1     0     1     2     0     0     0     0     0     0
5     2     0     1     0     0     2     0     0     0     0     1     0     0
6     1     0     0     1     0     0     0     0     2     0     1     0     0
  snp93 snp94 snp95 snp96 snp97 snp98 snp99 snp100
1     1     0     0     0     1     0     1      1
2     1     0     0     0     0     1     1      2
3     1     1     0     0     1     2     0      0
4     1     0     0     0     1     0     0      0
5     1     0     0     1     1     2     1      1
6     0     0     0     1     1     0     1      1
...

Covariates:
data frame with 0 columns and 0 rows


Results:

$Qstatistic
         Q1       Q2       Q3
1 0.1910558 71.34471 16.59228

$p.value
       p.Q1         p.Q2       p.Q3
1 0.6620394 1.293491e-05 0.00011429
\end{Soutput}
\end{Schunk}

\subsection{Preprocessing with Partial Least Square Discriminant Analysis (PLS-DA)}

This method of data preprocessing is used for dichotomous outcome.

\begin{Schunk}
\begin{Sinput}
> library(rqt)
> data <- read.table(system.file("extdata/test.bin1.dat",package="rqt"), 
+     header=TRUE)
> pheno <- data$pheno
> geno <- data[, 2:dim(data)[2]]
> obj <- rqtClass(phenotype=pheno, genotype=geno)
> res <- rQTest(obj, method="pls", out.type = "D", scale = TRUE)
\end{Sinput}
\begin{Soutput}
      R2X(cum) R2Y(cum) Q2(cum) RMSEE pre ort pR2Y pQ2
Total   0.0187    0.301  -0.292   0.2   1   0    1 0.7
\end{Soutput}
\begin{Sinput}
> print(res)
\end{Sinput}
\begin{Soutput}
Phenotype:
[1] 1 1 1 1 1 1
...

Genotype:
  snp1 snp2 snp3 snp4 snp5 snp6 snp7 snp8 snp9 snp10 snp11 snp12 snp13 snp14
1    0    0    0    1    0    2    1    0    0     0     2     1     2     0
2    1    0    1    0    0    1    0    0    0     0     0     0     1     0
3    0    0    0    0    1    0    0    1    0     1     1     1     0     0
4    0    0    1    0    1    0    0    1    1     0     0     0     1     0
5    0    0    1    1    1    1    1    1    0     1     0     1     0     0
6    0    0    1    1    1    0    0    1    0     1     1     0     1     0
  snp15 snp16 snp17 snp18 snp19 snp20 snp21 snp22 snp23 snp24 snp25 snp26 snp27
1     0     2     1     1     0     0     1     0     0     0     0     2     1
2     0     0     2     1     2     1     0     1     0     0     0     2     1
3     0     1     1     0     0     1     1     1     0     0     0     1     0
4     0     0     1     0     0     2     0     1     0     0     0     1     0
5     0     0     0     0     0     0     1     0     0     1     0     1     2
6     0     2     0     0     1     0     1     1     0     0     0     2     0
  snp28 snp29 snp30 snp31 snp32 snp33 snp34 snp35 snp36 snp37 snp38 snp39 snp40
1     0     1     0     0     0     0     0     0     1     1     0     0     1
2     1     2     0     0     0     1     2     1     2     0     1     1     0
3     0     0     0     0     0     0     1     1     1     2     2     0     0
4     2     0     0     0     1     2     2     0     1     1     1     1     0
5     0     0     0     0     0     1     1     0     0     1     1     0     0
6     0     0     0     0     0     1     0     1     1     0     0     2     0
  snp41 snp42 snp43 snp44 snp45 snp46 snp47 snp48 snp49 snp50 snp51 snp52 snp53
1     0     0     1     0     0     0     0     1     1     0     0     0     1
2     1     0     0     1     0     0     0     2     0     1     0     1     0
3     0     1     0     1     0     2     1     1     1     0     0     0     1
4     0     0     1     1     0     0     1     1     2     0     1     0     2
5     2     2     0     0     0     1     2     1     0     0     0     0     1
6     1     1     2     0     0     0     1     1     1     0     0     1     1
  snp54 snp55 snp56 snp57 snp58 snp59 snp60 snp61 snp62 snp63 snp64 snp65 snp66
1     1     1     1     1     0     0     0     0     1     1     0     1     1
2     2     1     0     1     0     0     0     1     0     1     2     2     1
3     2     1     0     0     0     0     0     0     0     1     0     0     1
4     1     1     0     0     0     0     2     1     0     1     0     0     2
5     1     1     1     0     0     0     2     1     1     0     0     1     1
6     0     1     1     0     0     0     0     0     0     0     1     2     1
  snp67 snp68 snp69 snp70 snp71 snp72 snp73 snp74 snp75 snp76 snp77 snp78 snp79
1     1     0     1     1     2     1     1     1     0     0     2     1     1
2     1     0     1     0     1     0     0     0     0     1     1     2     1
3     1     0     1     2     1     0     1     1     0     0     0     0     1
4     1     0     0     1     1     1     0     0     0     0     0     1     0
5     0     0     1     1     0     0     1     0     1     0     0     1     0
6     1     0     0     0     0     0     1     1     1     1     1     1     2
  snp80 snp81 snp82 snp83 snp84 snp85 snp86 snp87 snp88 snp89 snp90 snp91 snp92
1     1     0     1     1     0     1     1     0     0     0     1     0     1
2     1     2     0     0     0     1     2     0     2     1     0     0     2
3     1     0     0     0     0     2     0     0     2     0     0     0     0
4     1     2     0     1     0     1     2     0     0     0     0     0     0
5     2     0     1     0     0     2     0     0     0     0     1     0     0
6     1     0     0     1     0     0     0     0     2     0     1     0     0
  snp93 snp94 snp95 snp96 snp97 snp98 snp99 snp100
1     1     0     0     0     1     0     1      1
2     1     0     0     0     0     1     1      2
3     1     1     0     0     1     2     0      0
4     1     0     0     0     1     0     0      0
5     1     0     0     1     1     2     1      1
6     0     0     0     1     1     0     1      1
...

Covariates:
data frame with 0 columns and 0 rows


Results:

$Qstatistic
        Q1       Q2       Q3
1 11.12606 2.322913 11.12606

$p.value
          p.Q1         p.Q2         p.Q3
1 0.0008512339 0.0008512339 0.0008512339
\end{Soutput}
\end{Schunk}

\subsection{Using additional covariates}

Quite often, researchers want to supply not only genetic 
data but also specific covariates, 
representic some physiological parameters or environment 
(for example, to evaluate 
hyphoteses of gene-environment interactions). 
In such cases, the package \texttt{rqt} 
can accept additional covariates, in form of 
\texttt{N} by \texttt{K} matrix, as provided below:

\begin{Schunk}
\begin{Sinput}
> library(rqt)
> data <- read.table(system.file("extdata/test.bin1.dat",package="rqt"), 
+     header = TRUE)
> pheno <- data$pheno
> geno <- data[, 2:dim(data)[2]]
> covars <- read.table(system.file("extdata/test.cova1.dat",package="rqt"), 
+     header=TRUE)
> obj <- rqtClass(phenotype=pheno, genotype=geno, covariates = covars)
> res <- rQTest(obj, method="pca", out.type = "D")
> print(res)
\end{Sinput}
\begin{Soutput}
Phenotype:
[1] 1 1 1 1 1 1
...

Genotype:
  snp1 snp2 snp3 snp4 snp5 snp6 snp7 snp8 snp9 snp10 snp11 snp12 snp13 snp14
1    0    0    0    1    0    2    1    0    0     0     2     1     2     0
2    1    0    1    0    0    1    0    0    0     0     0     0     1     0
3    0    0    0    0    1    0    0    1    0     1     1     1     0     0
4    0    0    1    0    1    0    0    1    1     0     0     0     1     0
5    0    0    1    1    1    1    1    1    0     1     0     1     0     0
6    0    0    1    1    1    0    0    1    0     1     1     0     1     0
  snp15 snp16 snp17 snp18 snp19 snp20 snp21 snp22 snp23 snp24 snp25 snp26 snp27
1     0     2     1     1     0     0     1     0     0     0     0     2     1
2     0     0     2     1     2     1     0     1     0     0     0     2     1
3     0     1     1     0     0     1     1     1     0     0     0     1     0
4     0     0     1     0     0     2     0     1     0     0     0     1     0
5     0     0     0     0     0     0     1     0     0     1     0     1     2
6     0     2     0     0     1     0     1     1     0     0     0     2     0
  snp28 snp29 snp30 snp31 snp32 snp33 snp34 snp35 snp36 snp37 snp38 snp39 snp40
1     0     1     0     0     0     0     0     0     1     1     0     0     1
2     1     2     0     0     0     1     2     1     2     0     1     1     0
3     0     0     0     0     0     0     1     1     1     2     2     0     0
4     2     0     0     0     1     2     2     0     1     1     1     1     0
5     0     0     0     0     0     1     1     0     0     1     1     0     0
6     0     0     0     0     0     1     0     1     1     0     0     2     0
  snp41 snp42 snp43 snp44 snp45 snp46 snp47 snp48 snp49 snp50 snp51 snp52 snp53
1     0     0     1     0     0     0     0     1     1     0     0     0     1
2     1     0     0     1     0     0     0     2     0     1     0     1     0
3     0     1     0     1     0     2     1     1     1     0     0     0     1
4     0     0     1     1     0     0     1     1     2     0     1     0     2
5     2     2     0     0     0     1     2     1     0     0     0     0     1
6     1     1     2     0     0     0     1     1     1     0     0     1     1
  snp54 snp55 snp56 snp57 snp58 snp59 snp60 snp61 snp62 snp63 snp64 snp65 snp66
1     1     1     1     1     0     0     0     0     1     1     0     1     1
2     2     1     0     1     0     0     0     1     0     1     2     2     1
3     2     1     0     0     0     0     0     0     0     1     0     0     1
4     1     1     0     0     0     0     2     1     0     1     0     0     2
5     1     1     1     0     0     0     2     1     1     0     0     1     1
6     0     1     1     0     0     0     0     0     0     0     1     2     1
  snp67 snp68 snp69 snp70 snp71 snp72 snp73 snp74 snp75 snp76 snp77 snp78 snp79
1     1     0     1     1     2     1     1     1     0     0     2     1     1
2     1     0     1     0     1     0     0     0     0     1     1     2     1
3     1     0     1     2     1     0     1     1     0     0     0     0     1
4     1     0     0     1     1     1     0     0     0     0     0     1     0
5     0     0     1     1     0     0     1     0     1     0     0     1     0
6     1     0     0     0     0     0     1     1     1     1     1     1     2
  snp80 snp81 snp82 snp83 snp84 snp85 snp86 snp87 snp88 snp89 snp90 snp91 snp92
1     1     0     1     1     0     1     1     0     0     0     1     0     1
2     1     2     0     0     0     1     2     0     2     1     0     0     2
3     1     0     0     0     0     2     0     0     2     0     0     0     0
4     1     2     0     1     0     1     2     0     0     0     0     0     0
5     2     0     1     0     0     2     0     0     0     0     1     0     0
6     1     0     0     1     0     0     0     0     2     0     1     0     0
  snp93 snp94 snp95 snp96 snp97 snp98 snp99 snp100
1     1     0     0     0     1     0     1      1
2     1     0     0     0     0     1     1      2
3     1     1     0     0     1     2     0      0
4     1     0     0     0     1     0     0      0
5     1     0     0     1     1     2     1      1
6     0     0     0     1     1     0     1      1
...

Covariates:
          COV1
1 -0.612463927
2 -0.464158885
3  0.006153597
4 -0.732109468
5 -0.223530136
6 -0.744903822


Results:

$Qstatistic
        Q1       Q2       Q3
1 2.012625 3.761859 2.012625

$p.value
       p.Q1      p.Q2      p.Q3
1 0.1559952 0.8258796 0.2895166
\end{Soutput}
\end{Schunk}


\section{Meta-analysis}

\begin{Schunk}
\begin{Sinput}
> library(rqt)
> data1 <- read.table(system.file("extdata/phengen2.dat",package="rqt"), skip=1)
> obj1 <- rqtClass(phenotype=data1[,1], genotype=data1[, 2:dim(data1)[2]])
> data2 <- read.table(system.file("extdata/phengen3.dat",package="rqt"), skip=1)
> obj2 <- rqtClass(phenotype=data2[,1], genotype=data2[, 2:dim(data2)[2]])
> data3 <- read.table(system.file("extdata/phengen.dat",package="rqt"), skip=1)
> obj3 <- rqtClass(phenotype=data3[,1], genotype=data3[, 2:dim(data3)[2]])
> res.meta <- rQTestMeta(list(obj1, obj2, obj3))
> print(res.meta)
\end{Sinput}
\begin{Soutput}
$final.pvalue
[1] 0.004276623

$pvalueList
[1] 0.000858092 0.367634896 0.245240026

$df
[1] 6

$chi.comb
[1] 18.93396
\end{Soutput}
\end{Schunk}


\section{Session information}
\begin{Schunk}
\begin{Sinput}
> sessionInfo()
\end{Sinput}
\begin{Soutput}
R version 3.2.4 (2016-03-10)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X 10.12.1 (unknown)

locale:
[1] C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] rqt_0.99.0     ropls_1.2.14   plyr_1.8.4     pls_2.5-0      glmnet_2.0-5  
[6] foreach_1.4.3  Matrix_1.2-7.1

loaded via a namespace (and not attached):
[1] CCP_1.1            parallel_3.2.4     tools_3.2.4        Rcpp_0.12.8       
[5] codetools_0.2-15   grid_3.2.4         iterators_1.0.8    CompQuadForm_1.4.1
[9] lattice_0.20-34   
\end{Soutput}
\end{Schunk}
%--------
\clearpage
\section{References}
\begin{thebibliography}{00}

\bibitem{Wu2011} Wu, M. C., Lee, S., Cai, T., Li, Y., Boehnke, M., 
and Lin, X. (2011) 
Rare Variant Association Testing for Sequencing Data Using the Sequence 
Kernel Associa-tion Test (SKAT). The American Journal of 
Human Genetics, 89, 82-93.
\bibitem{Li2008} Li, B. and Leal, S.M. (2008) Methods for 
Detecting Associations 
with Rare Vari-ants for Common Diseases: Application to Analysis 
of Sequence Data. The American Journal of Human Genetics, 83, 311-321.
\bibitem{Liu2010} Liu D., Leal S. (2010) A Novel Adaptive Method for 
the Analysis 
of Next-Generation Sequencing Data to Detect Complex Trait 
Associations with Rare Variants Due to Gene Main Effects and 
Interactions, PLoS Genet., 6(10).
\bibitem{Madsen2009} Madsen, B.E, Browning, S.R. (2009) 
A Groupwise Association 
Test for Rare Mutations Using a Weighted Sum Statistic, PLoS Genet., 5(2).

\bibitem{Lee2016} Lee, J., Kim, Y.J., Lee, J., T2D-Genes Consortium, 
Kim, B-J., 
Lee, S., Park T. (2016) Gene-set association tests for 
next-generation sequencing data, Bioinformatics, 32(17).

\bibitem{Tibshirani1996} Tibshirani, R. (1996) Regression shrinkage 
and selection 
via the lasso. J. Royal. Statist. Soc B., Vol. 58, No. 1, pages 267-288.
\end{thebibliography}
%--------

\end{document}
